runtime/src/gpu/common/cuda-shared.h

```diff
@@ -28,16 +28,21 @@
 #include "cuda-utils.h"
 
 static inline
-void* chpl_gpu_getKernel(const char* fatbinData, const char* kernelName) {
+void* chpl_gpu_load_module(const char* fatbin_data) {
+  CUmodule cuda_module;
 
-  CUmodule    cudaModule;
-  CUfunction  function;
+  CUDA_CALL(cuModuleLoadData(&cuda_module, fatbin_data));
+  assert(cuda_module);
 
-  // Create module for object
-  CUDA_CALL(cuModuleLoadData(&cudaModule, fatbinData));
+  return (void*)cuda_module;
+}
+
+static inline
+void* chpl_gpu_load_function(CUmodule cuda_module, const char* kernel_name) {
```

Comment:
I think I'd just have this function call `chpl_gpu_cuda_modules[chpl_task_getRequestedSubloc()]` directly rather than passing it in as a parameter.

I'm also wondering how optimized the `cuModuleGetFunction` itself is. Like does it do a bunch of string comparisons internally or does it have some kind of hashed lookup table?

Anyway not something I need answer for, for this PR, just brainstorming potential places for further optimization.

Reply:
> I think I'd just have this function call chpl_gpu_cuda_modules[chpl_task_getRequestedSubloc()] directly rather than passing it in as a parameter.

I agree.

However, currently these are in the shared header. Having this function access `chpl_gpu_cuda_modules` would mean moving that array to the header, which I want to avoid.

The reason we have this header is because we wanted to call these functions from libomptarget layer, if/when we have it. In the grand scheme of things, I hope that this header will disappear soon, the functions will move to `gpu-cuda.c`, and then we can more easily do as you suggest.

Do you insist we have a solution with this PR? Or are you OK with me moving on as-is?

---

> I'm also wondering how optimized the cuModuleGetFunction itself is. Like does it do a bunch of string comparisons internally or does it have some kind of hashed lookup table?

Very :) My initial implementation of this was epitome of premature optimization. I had a function table where I cache the loaded functions once they launch. I thought the bottleneck was function load and not the module load. (The branch is named `gpu-cached-kernels`, even). But, profiling, I noticed that what I am earning by caching functions was next to nothing. And I wasn't making anything parallel-safe at that moment.

After I noticed all this, looking a little closer at the documentation, I found out about the `CUDA_MODULE_LOADING` variable which

> Specifies the module loading mode for the application. When set to EAGER, all kernels from a cubin, fatbin or a PTX file are fully loaded upon corresponding cuModuleLoad* API call. This is the same behavior as in all preceding CUDA releases. When set to LAZY, loading of a specific kernel is delayed to the point a CUfunc handle is extracted with cuModuleGetFunction API call. This mode allows for lowering initial module loading latency and decreasing initial module-related device memory consumption, at the cost of higher latency of cuModuleGetFunction API call. Default behavior is EAGER. Default behavior may change in future CUDA releases.

https://docs.nvidia.com/cuda/cuda-c-programming-guide/

Apparently, you can query that env using `cuModuleGetLoadingMode` programmatically.

Reply:
> Do you insist we have a solution with this PR? Or are you OK with me moving on as-is?

Sure, I was making this more as a style preference than correctness issue so if you think the knock-on effects are even worse I'd keep it as-is.

> After I noticed all this, looking a little closer at the documentation, I found out about the CUDA_MODULE_LOADING ...

That's interesting to know about. I wonder if larger applications benefit from using the `LAZY` scheme since I imagine a lot of programs will have kernels that never get called (well, at least until we write an analysis identifying what needs to be gpuized).

Reply:
Right. We might even want to allow (power-)users to change the behavior for faster application startup. Something to keep in mind for sure.